enhancements/subscription-content/subscription-injection.md

```diff
@@ -0,0 +1,654 @@
+  yaml: |
+    apiVersion: subscription.openshift.io/v1
+    kind: ClusterBundle
+    metadata:
+      name: cluster
+    spec:
+      aggregateToClusterRoles:
+      - edit
+      - admin
+      entitlements:
+      - name: etc-pki-entitlement
+        namespace: openshift-config
+```
+
+```yaml
+apiVersion: console.openshift.io/v1
+kind: ConsoleYAMLSample
+metadata:
+  name: subscription-namespaced-bundle-example
+spec:
+  title: Namespaced subscription bundle example
+  description: |
+    An example of a subscription bundle that can be injected into any workload in the bundle's
+    namespace that has the `subscription.openshift.io/inject-bundle: subscription` annotation in
+    its pod spec.
+  targetResource:
+    apiVersion: subscription.openshift.io/v1
+    kind: Bundle
+  yaml: |
+    apiVersion: subscription.openshift.io/v1
+    kind: Bundle
+    metadata:
+      name: subscription
+    spec:
+      entitlements:
+      - name: etc-pki-entitlement
+```
+
+The `entitlements` object is a reference to the `Secrets` which contain the entitlement keys for
+the cluster's subscription. The cluster administrator is responsible for obtaining the entitlement
+keys and adding them to the cluster. Administrators can refer to existing guidance on how to obtain
+this information.
+
+The API specs will also include a `yumRepositories` field, which allows for a list of `ConfigMaps`
+containing `.repo` configuration files that allow an entitlement to be downloaded from a Red Hat
+Satellite instance.
+
+#### Subscription Cluster Bundle Controller
+
+When a `ClusterBundle` object is created, the Subscription Bundle Controller (a new controller)
+will perform the necessary actions to make the bundle available to cluster workloads. This consists
+of the following:
+
+1. Verify that the referenced `entitlements` and optional `yumRepositories` resources exist.
+2. Create the ProjectedResource CSI driver `Shares` for the `entitlements` and (optional)
+  `yumRepositories` as declared in the Projected Resource CSI Driver proposal.
+3. Create a `ClusterRole` associated with the bundle, which grants `get`, `watch`, and `list`
+   permissions to the associated `Shares` and the `ClusterBundle` object.
+4. Aggregate the cluster role to the provided list of system roles.
+5. Add owner references to the generated `ClusterRole` and shares. This ensures the dependent
+   objects are deleted when the `ClusterBundle` is deleted.
+
+When a `ClusterBundle` object is updated, changes to the referenced `entitlements` and
+`yumRepositories` are propagated to the associated `Share` objects. Likewise, changes to aggregated
+cluster roles should be propagated to the associated `ClusterRole`. Once completed, non-terminated
+pods which inject the `ClusterBundle` and whose recorded generation does not match the
+`ClusterBundle`'s new generation need to be evicted. This will allow new pods to be created with
+appropriate volume mounts for the cluster bundles.
```

Comment:
i think we need to discuss this some more.  That's a pretty hard constrraint, so i want to talk about what else we can/can't do.  And perhaps, if it's really bad, the better answer is to say "you can't edit cluster bundles, you have to delete and recreate them to change them".  (still not sure what that means for a running pod w/ a ref to the clusterbundle in question though).

but in general, let's think about whether we need to put more explicit constraints on expectations, rather than a "it kinda works but only if you do this thing that no one ever wants to do"

Reply:
For `Bundle` objects, the only changes to `spec` that are allowed are adds/removals from the `entitlements` and `yumRepositories` lists. Propagating this to pods requires changing the `Volume`, which can't be done on a running pod.
We can block edits to `spec` via a separate validating webhook.
For deleted `Bundles`, a potentially valid alternative is to leave the pods as they are.

With `ClusterBundle`, we have the added complexity of the cluster aggregation roles. We probably need a similar/same approach to edits - block edits to `spec` via a webhook. And likewise for deletions, rather than using OwnerRefs we instead track our related objects in `status` and leave them in place on deletion. The only thing we may do is change the referenced cluster role so that it no longer grants read permission on the removed `ClusterBundle` object.

Reply:
on further thinking about this, this may not be as big a deal as i was thinking, since i expect 99% of the usage of this feature to be "launch pod, immediately use entitlements to pull in content". 

I suppose someone might try to have a long-lived builder pod that provides some sort of build service where it needs to periodically grab new rpm content, but that seems unlikely.

So it might be reasonable to just assert up front that the CSI driver is based on a "you get the content based on the state at pod launch and there are no dynamic updates to the content" model.

it needs to be stated very clearly and it's possible there will be some pushback, but given how much it simplifies things and the fact that i think it achieves the primary use case here, i'd be ok w/ it personally.

Reply:
I've attempted to clarify the behavior in a few places:

1. In "Non-goals", I've made automatic key rotation and "add an extra entitlement via a new `Secret`" out of scope.
2. Updated the `Bundle` and `ClusterBundle` behavior to keep non-terminated pods in place on update and delete.
3. Add the risk of "Containers run with outdated entitlement keys" - with the mitigation of "update the secret with the original keys".